ruby-ll won’t compile with Ruby 3.4

Since switching to Ruby 3.4, the native extension part of ruby-ll won’t compile anymore.

Here’s the result of `gem install ruby-ll --verbose`

```
HEAD https://index.rubygems.org/versions
200 OK
GET https://index.rubygems.org/info/ruby-ll
200 OK
GET https://index.rubygems.org/info/ansi
200 OK
GET https://index.rubygems.org/info/ast
200 OK
Downloading gem ruby-ll-2.1.3.gem
GET https://index.rubygems.org/gems/ruby-ll-2.1.3.gem
Fetching ruby-ll-2.1.3.gem
200 OK
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/.yardopts
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/LICENSE
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/README.md
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/bin/ruby-ll
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/doc/css/common.css
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/doc/driver_architecture.md
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/c/driver.c
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/c/driver.h
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/c/driver_config.c
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/c/driver_config.h
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/c/extconf.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/c/khash.h
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/c/kvec.h
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/c/libll.c
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/c/libll.h
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/c/macros.h
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/java/Libll.java
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/java/org/libll/Driver.java
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/java/org/libll/DriverConfig.java
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/ast/node.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/branch.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/cli.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/code_generator.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/compiled_configuration.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/compiled_grammar.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/configuration_compiler.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/driver.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/driver_config.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/driver_template.erb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/epsilon.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/erb_context.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/grammar_compiler.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/lexer.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/message.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/operator.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/parser.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/parser_error.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/rule.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/setup.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/source_line.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/terminal.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/token.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/lib/ll/version.rb
/home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ruby-ll.gemspec
Building native extensions. This could take a while...
current directory: /home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/c
["/usr/bin/ruby", "extconf.rb"]
creating Makefile
current directory: /home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/c
["make", "DESTDIR=", "sitearchdir=./.gem.20250520-25244-oxwwcg", "sitelibdir=./.gem.20250520-25244-oxwwcg", "clean"]
current directory: /home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/c
["make", "DESTDIR=", "sitearchdir=./.gem.20250520-25244-oxwwcg", "sitelibdir=./.gem.20250520-25244-oxwwcg"]
compiling driver.c
driver.c: Dans la fonction « ll_driver_allocate »:
driver.c:54:5: attention: « rb_data_object_get_warning » est obsolète: by TypedData [-Wdeprecated-declarations]
   54 |     Data_Get_Struct(config, DriverConfig, state->config);
      |     ^~~~~~~~~~~~~~~
Dans le fichier inclus depuis /usr/include/ruby-3.4.0/ruby/internal/core.h:27,
                 depuis /usr/include/ruby-3.4.0/ruby/ruby.h:29,
                 depuis /usr/include/ruby-3.4.0/ruby.h:38,
                 depuis driver.h:4,
                 depuis driver.c:1:
/usr/include/ruby-3.4.0/ruby/internal/core/rdata.h:325:1: note: déclaré ici
  325 | rb_data_object_get_warning(VALUE obj)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~
driver.c:59:5: attention: « rb_data_object_wrap_warning » est obsolète: by TypedData [-Wdeprecated-declarations]
   59 |     return Data_Wrap_Struct(klass, ll_driver_mark, ll_driver_free, state);
      |     ^~~~~~
/usr/include/ruby-3.4.0/ruby/internal/core/rdata.h:293:1: note: déclaré ici
  293 | rb_data_object_wrap_warning(VALUE klass, void *ptr, RUBY_DATA_FUNC mark, RUBY_DATA_FUNC free)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~
driver.c: Dans la fonction « ll_driver_each_token »:
driver.c:92:5: attention: « rb_data_object_get_warning » est obsolète: by TypedData [-Wdeprecated-declarations]
   92 |     Data_Get_Struct(self, DriverState, state);
      |     ^~~~~~~~~~~~~~~
/usr/include/ruby-3.4.0/ruby/internal/core/rdata.h:325:1: note: déclaré ici
  325 | rb_data_object_get_warning(VALUE obj)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~
driver.c: Dans la fonction « ll_driver_parse »:
driver.c:312:5: attention: « rb_data_object_get_warning » est obsolète: by TypedData [-Wdeprecated-declarations]
  312 |     Data_Get_Struct(self, DriverState, state);
      |     ^~~~~~~~~~~~~~~
/usr/include/ruby-3.4.0/ruby/internal/core/rdata.h:325:1: note: déclaré ici
  325 | rb_data_object_get_warning(VALUE obj)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~
Dans le fichier inclus depuis /usr/include/ruby-3.4.0/ruby/internal/assume.h:29,
                 depuis /usr/include/ruby-3.4.0/ruby/backward/2/assume.h:24,
                 depuis /usr/include/ruby-3.4.0/ruby/defines.h:72,
                 depuis /usr/include/ruby-3.4.0/ruby/ruby.h:25:
/usr/include/ruby-3.4.0/ruby/internal/anyargs.h:363:45: erreur: passage de l'argument 5 de « rb_block_call » depuis un type pointeur incompatible [-Wincompatible-pointer-types]
  363 | # define RUBY_METHOD_FUNC(func) RBIMPL_CAST((VALUE (*)(ANYARGS))(func))
/usr/include/ruby-3.4.0/ruby/internal/cast.h:31:29: note: dans la définition de la macro « RBIMPL_CAST »
   31 | # define RBIMPL_CAST(expr) (expr)
      |                             ^~~~
driver.c:329:9: note: dans l'expansion de la macro « RUBY_METHOD_FUNC »
  329 |         RUBY_METHOD_FUNC(ll_driver_each_token),
      |         ^~~~~~~~~~~~~~~~
Dans le fichier inclus depuis /usr/include/ruby-3.4.0/ruby/ruby.h:41:
/usr/include/ruby-3.4.0/ruby/internal/iterator.h:324:90: note: « rb_block_call_func_t » {alias « long unsigned int (*)(long unsigned int,  long unsigned int,  int,  const long unsigned int *, long unsigned int) »} attendu mais l'argument est de type « VALUE (*)(void) » {alias « long unsigned int (*)(void) »}
  324 | VALUE rb_block_call(VALUE obj, ID mid, int argc, const VALUE *argv, rb_block_call_func_t proc, VALUE data2);
      |                                                                     ~~~~~~~~~~~~~~~~~~~~~^~~~
/usr/include/ruby-3.4.0/ruby/internal/iterator.h:88:29: note: « rb_block_call_func_t » déclaré ici
   88 | typedef rb_block_call_func *rb_block_call_func_t;
      |                             ^~~~~~~~~~~~~~~~~~~~
make: *** [Makefile:251: driver.o] Error 1
ERROR:  Error installing ruby-ll:
	ERROR: Failed to build gem native extension.

    current directory: /home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/c
/usr/bin/ruby extconf.rb
current directory: /home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/c
make DESTDIR\= sitearchdir\=./.gem.20250520-25244-oxwwcg sitelibdir\=./.gem.20250520-25244-oxwwcg clean
current directory: /home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3/ext/c
make DESTDIR\= sitearchdir\=./.gem.20250520-25244-oxwwcg sitelibdir\=./.gem.20250520-25244-oxwwcg
Building has failed. See above output for more information on the failure.
make failed, exit code 2

Gem files will remain installed in /home/myname/.local/share/gem/ruby/3.4.0/gems/ruby-ll-2.1.3 for inspection.
Results logged to /home/myname/.local/share/gem/ruby/3.4.0/extensions/x86_64-linux/3.4.0/ruby-ll-2.1.3/gem_make.out
```

Sorry it’s in French. Most of it is just warnings, but this in the error.

```
/usr/include/ruby-3.4.0/ruby/internal/anyargs.h:363:45: erreur: passage de l'argument 5 de « rb_block_call » depuis un type pointeur incompatible [-Wincompatible-pointer-types]
  363 | # define RUBY_METHOD_FUNC(func) RBIMPL_CAST((VALUE (*)(ANYARGS))(func))
```

Meaning argument 5 of `rb_block_call` gets passed using an incompatible pointer type.

Here’s for `gem env`

```
RubyGems Environment:
  - RUBYGEMS VERSION: 3.6.9
  - RUBY VERSION: 3.4.3 (2025-04-14 patchlevel 32) [x86_64-linux]
  - INSTALLATION DIRECTORY: /usr/lib/ruby/gems/3.4.0
  - USER INSTALLATION DIRECTORY: /home/myname/.local/share/gem/ruby/3.4.0
  - CREDENTIALS FILE: /home/myname/.local/share/gem/credentials
  - RUBY EXECUTABLE: /usr/bin/ruby
  - GIT EXECUTABLE: /usr/bin/git
  - EXECUTABLE DIRECTORY: /usr/bin
  - SPEC CACHE DIRECTORY: /home/myname/.cache/gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /etc
  - RUBYGEMS PLATFORMS:
     - ruby
     - x86_64-linux
  - GEM PATHS:
     - /usr/lib/ruby/gems/3.4.0
     - /home/myname/.local/share/gem/ruby/3.4.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => true
     - :bulk_threshold => 1000
     - "gem" => "--user-install"
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /bin/internet
     - /bin/media
     - /bin/mensaria
     - /home/myname/.local/bin
     - /usr/local/bin
     - /usr/bin
     - /bin
     - /usr/local/sbin
     - /usr/lib/jvm/default/bin
     - /usr/bin/site_perl
     - /usr/bin/vendor_perl
     - /usr/bin/core_perl
```

Obviously, I’m using the latest version of RubyGems.

--- a/ext/c/driver.c
+++ b/ext/c/driver.c
@@ -66,7 +66,13 @@ VALUE ll_driver_allocate(VALUE klass)
  * @param token An Array containing the token type as a Symbol and its value.
  * @param self The Driver instance currently in use.
  */
-VALUE ll_driver_each_token(VALUE token, VALUE self)
+VALUE ll_driver_each_token(
+    VALUE token,
+    VALUE self,
+    int argc,
+    const VALUE *argv,
+    VALUE block_arg
+)
 {
     VALUE method;
     VALUE action_args;
@@ -326,7 +332,7 @@ VALUE ll_driver_parse(VALUE self)
         id_each_token,
         0,
         NULL,
-        RUBY_METHOD_FUNC(ll_driver_each_token),
+        ll_driver_each_token,
         self
     );
 

--- a/ext/c/driver.c
+++ b/ext/c/driver.c
@@ -60,11 +60,16 @@ VALUE ll_driver_allocate(VALUE klass)
 }
 
 /**
- * Callback function for iterating over every input token and actually parsing
- * said input.
+ * Callback function for iterating over every input token and parsing it.
+ *
+ * This function is intended to be used as a block callback for Ruby's
+ * `rb_block_call`, and processes a single token per invocation.
  *
  * @param token An Array containing the token type as a Symbol and its value.
  * @param self The Driver instance currently in use.
+ * @param argc The number of additional arguments passed to the block.
+ * @param argv The array of additional arguments passed to the block.
+ * @param block_arg The block argument passed from Ruby (if any).
  */
 VALUE ll_driver_each_token(
     VALUE token,

